runtime/frame.go

```diff
@@ -39,20 +39,43 @@ type Frame struct {
 	globals     *Dict `attr:"f_globals"`
 	lineno      int   `attr:"f_lineno"`
 	code        *Code `attr:"f_code"`
+	taken       bool
 }
 
 // NewRootFrame creates a Frame that is the bottom of a new stack.
 func NewRootFrame() *Frame {
-	return newFrame(nil)
+	f := &Frame{Object: Object{typ: FrameType}}
+	f.pushFrame(nil)
+	return f
 }
 
 // newFrame creates a new Frame whose parent frame is back.
 func newFrame(back *Frame) *Frame {
-	f := &Frame{Object: Object{typ: FrameType}}
+	f := back.frameCache
```

Comment:
So long as the field being referenced is unambiguous, this is fine. When the reference is ambiguous, it is a compile error: https://play.golang.org/p/XbdM_FTH4b In the spec they are called "promoted" fields (and methods).